zholobov.oleg/S2

OlegZholobov opened the pull request:
![image](https://github.com/user-attachments/assets/fc43b521-7b13-4429-a8b0-4feaed264d83)

OlegZholobov committed:
Initial commit for S2

added minimal main.cpp
added Stack and Queue interfaces

OlegZholobov committed:
wrote tests for Queue and Stack

they all fail until Queue and Stack will be implemented

OlegZholobov committed:
enhanced interface

fixed tests

OlegZholobov committed:
implemented Stack

added some checks in tests

OlegZholobov committed:
implemented Queue

added some checks in tests

OlegZholobov committed:
implemented S2

hopefully it will pass acceptance

OlegZholobov committed:
fixed overflows for all operations

OlegZholobov committed:
added test for clear() method in Stack and Queue

fixed error case in clear() in Stack and Queue

Volgarenok commented on zholobov.oleg/S2/main.cpp:
```diff
@@ -0,0 +1,252 @@
+      }
+    }
+    return true;
+  }
```
Не должно быть функции для проверки, является ли токен/строка числом. Там, где вы ожидаете число, преобразуйте строку соответствующими функциями стандартной библиотеки.

OlegZholobov replied:
сделал функцию strToLong для этих целей, использует std::strtol()

Volgarenok replied:
Так вам вообще не нужна она. В том месте. Вам не надо перебирать эту альтернативу. Просто всё, что не является скобкой/операцией должно быть числом. Вам достаточно strtol, возвращающей результат

Volgarenok replied:
Не убрали

Volgarenok replied:
Лично я - не опаздываю. Вы - не знаю

OlegZholobov replied:
тогда случайно не заметил не закрытый комментарий. Вчера его исправил в коммите [85e60d0](https://github.com/Volgarenok/spbspu-labs-2025-aads-a/pull/92/commits/85e60d0c61e30234c709937c1f251e4d8a6828bc) так как вы советовали, то есть всё что не скобка и не операция - это число. Я не понял, ещё что-то исправить надо?

Volgarenok commented on zholobov.oleg/S2/main.cpp:
```diff
@@ -0,0 +1,252 @@
+    return (token == "+") || (token == "-") || (token == "*") || (token == "/") || (token == "%");
+  }
+
+  int operatorPriority(const std::string& op)
```
Вы не должны предоставлять пользователям конкретные значения приоритетов - это деталь реализации. Вместо этого вы должны предоставить для них функцию, выполняющую сравнение

OlegZholobov replied:
переделал. Теперь пользователь видит только bool isLessPriority(op1, op2)

Volgarenok commented on zholobov.oleg/S2/main.cpp:
```diff
@@ -0,0 +1,252 @@
+      } else if (isOperator(token)) {
+        while ((!op_stack.empty()) &&
+               (op_stack.top() != "(") &&
+               (operatorPriority(op_stack.top()) >= operatorPriority(token))) {
```
Что это за треш?

Volgarenok commented on zholobov.oleg/S2/main.cpp:
```diff
@@ -0,0 +1,252 @@
+    }
+    return result;
+  }
+
```
Что это за универсальный всемогутор?

OlegZholobov replied:
провел рефакторинг. Вынес определение переполнений для всех операций в отдельные функции. Вроде стало получше.

Volgarenok commented on zholobov.oleg/S2/main.cpp:
```diff
@@ -0,0 +1,252 @@
+  Queue< std::string > stringToTokens(const std::string& str)
+  {
+    Queue< std::string > tokens;
+    std::istringstream iss(str);
```
Вы задание открывали в принципе?

OlegZholobov replied:
Убрал использование того что не положено )

Volgarenok commented on zholobov.oleg/S2/queue.hpp:
```diff
@@ -0,0 +1,205 @@
+
+template < typename T >
+zholobov::Queue< T >::Queue(const Queue& other):
+  data_(nullptr), capacity_(0), head_(0), size_(0)
```
CG

Volgarenok commented on zholobov.oleg/S2/queue.hpp:
```diff
@@ -0,0 +1,205 @@
+{
+  push(T(value));
+}
+
```
А зачем ещё отдельно копирование делать?

OlegZholobov replied:
Этот метод принимает const T& value и должен положить в контейнер копию value. Вообще его реализация такая же как и push(T&& value). Чтобы не дублировать код вызываю push(T&& value) с копией значения.
Временный объект T(value) - это копия value. Тип у этого временного объекта: T&&. Так что, все кажется  нормально.

Для улучшения читаемости кода, я добавил там std::move, чтобы явно указать что он T&&
То есть исправил на push(std::move(T(value)));

OlegZholobov replied:
Я ещё раз подумал и решил, что просто одно копирование в push_back(const T&) будет лучше, чем копирование при создании временного объекта, а потом его перемещение в методе push_back(T&&).
В общем, переделал. Общий код вынес в отдельный приватный метод realloc(). Вроде стало получше.

Volgarenok commented on zholobov.oleg/S2/queue.hpp:
```diff
@@ -0,0 +1,205 @@
+  }
+  size_t tail = (head_ + size_ - 1) % capacity_;
+  return *data_[tail];
+}
```
У очереди только `front`

Volgarenok commented on zholobov.oleg/S2/queue.hpp:
```diff
@@ -0,0 +1,205 @@
+    void swap(Queue& other) noexcept;
+
+  private:
+    T** data_;
```
Ну нет. В очереди должны лежать данные, а не указатели на них. Зачем лишняя косвенность?

OlegZholobov replied:
Казалось что сами данные лучше хранить отдельно, создавая и удаляя и тем самым экономить память.
Но если подумать, то есть минусы - много лишних new/delete.
Так что убрал эту косвенность. Теперь всё хранится прямо в массиве.

Volgarenok commented on zholobov.oleg/S2/stack.hpp:
```diff
@@ -0,0 +1,169 @@
+      for (; size_ < other.size_; ++size_) {
+        data_[size_] = new T(*other.data_[size_]);
+      }
+    } catch (const std::bad_alloc&) {
```
А если не bad_alloc?

OlegZholobov replied:
исправил

Volgarenok commented on zholobov.oleg/S2/stack.hpp:
```diff
@@ -0,0 +1,169 @@
+namespace zholobov {
+
+  template < typename T >
+  class Stack {
```
Много дублирований с очередью. Нехорошо

OlegZholobov replied:
Сделал отдельный класс Array (динамический массив, почти как предыдущий Queue), с возможностью добавлять элементы в конец и брать/убирать элементы из начала и конца. На его основе переделал Queue и Stack. Т.е. они теперь адаптеры над Array.

Volgarenok commented on zholobov.oleg/S2/test-queue.cpp:
```diff
@@ -0,0 +1,171 @@
+  }
+  for (int i = 0; i < 48; ++i) {
+    queue.pop();
+  }
```
Вам надо минимизировать размер тестов. В том числе по объёму выполняемого кода. Ваша задача выделить такой минимальный объём, который действительно проверяет что-то

OlegZholobov replied:
Этот тест предназначен для проверки правильности работы механизма увеличения размера динамического массива. Я бы его лучше оставил.
А теперь, когда появился класс Array, этот тест будет для него. А из тестов для Queue и Stack я его убрал, он там не нужен.

OlegZholobov committed:
changes per comments

OlegZholobov committed:
Merge remote-tracking branch 'volgarenok/master' into zholobov.oleg/S2

OlegZholobov committed:
Merge branch 'zholobov.oleg/S1' into zholobov.oleg/S2

OlegZholobov committed:
added range check after std::strtol()

OlegZholobov commented:
Tests PASSED
https://github.com/OlegZholobov/spbspu-labs-2025-aads-a/actions/runs/15096062282

Volgarenok commented:
@OlegZholobov Если S1 принята, все тесты должны проходить здесь

OlegZholobov committed:
Merge remote-tracking branch 'volgarenok/master' into zholobov.oleg/S2

OlegZholobov committed:
small optimizations

- added realloc() for storage extention
- optimized push_back(const T&) and push_back(T&&)
- typo in exceptions explanatory string

OlegZholobov committed:
moved array, queue, stack to common dir

OlegZholobov committed:
Merge remote-tracking branch 'volgarenok/master' into zholobov.oleg/S2

OlegZholobov committed:
very minor fix

Volgarenok commented on zholobov.oleg/common/queue.hpp:
```diff
@@ -0,0 +1,26 @@
+    void pop() { container_.pop_front(); }
+    bool empty() const { return container_.empty(); }
+    void clear() noexcept { container_.clear(); }
+    void swap(Queue& other) noexcept { container_.swap(other); };
```
Ну вы издеваетесь? Как я тут читать это должен?

OlegZholobov replied:
так просто намного меньше места занимало, поэтому решил так попробовать сделать. 
Привёл в соответствие с CG

Volgarenok commented on zholobov.oleg/common/queue.hpp:
```diff
@@ -0,0 +1,26 @@
+  class Queue {
+  public:
+    void push(const T& value) { container_.push_back(value); }
+    void push(T&& value) { container_.push_back(std::forward< T >(value)); }
```
Это не универсальная ссылка, вам не нужен forward

OlegZholobov replied:
переделал как надо

Volgarenok commented on zholobov.oleg/common/array.hpp:
```diff
@@ -0,0 +1,234 @@
+  } catch (...) {
+    delete[] new_data;
+    throw;
+  }
```
Проблема в том, что при перемещении старых объектов у вас данные окажутся в странном состоянии. Вы не будете знать, сколько объектов вы успели переместить

OlegZholobov replied:
похоже по нормальному это можно решить с использованием неинициализированного массива и placement new. Надеюсь правильно переделал

Volgarenok replied:
Какую-то фигню вы сделали. Этой задаче сто лет в обед. Мы делали её в том семестре: задача расширения существующего массива. Я не понимаю зачем вы сложностей себе каких-то нагородили. Верните как было

Volgarenok replied:
Ну ладно, тут работает

OlegZholobov replied:
у меня же предыдущая версия не во всех случая правильно работала. В текущей версии все недочёты исправил, плюс теперь нет требования к классу T иметь конструктор по умолчанию как было раньше и если у него есть noexcept-move-constructor, то ещё и при realloc массива он будет использоваться для перемещения элементов. Мне кажется так лучше оставить

Volgarenok commented on zholobov.oleg/common/array.hpp:
```diff
@@ -0,0 +1,234 @@
+    T& back();
+    void pop_front();
+    void pop_back();
+    bool empty() const;
```
Ну empty наверное тоже noexcept

OlegZholobov replied:
забыл, исправил

OlegZholobov committed:
fixes per comments

OlegZholobov committed:
minor fix

OlegZholobov committed:
Merge remote-tracking branch 'volgarenok/master' into zholobov.oleg/S2

OlegZholobov committed:
fix infixToPostfix

Assume a token is a number unless it is one of parenthesis or operator

OlegZholobov committed:
Merge remote-tracking branch 'volgarenok/master' into zholobov.oleg/S2

Volgarenok commented:
@OlegZholobov Я вас просил убрать эту функцию в принципе, она не нужна. Она делает то же самое, что и стандартная практически

Вы её использовали неправильно: хотя бы в части нейтральности относительно исключений
